src/oscar/apps/customer/utils.py

```diff
@@ -24,65 +24,76 @@ def __init__(self, logger=None):
 
     def dispatch_direct_messages(self, recipient, messages):
         """
-        Dispatch one-off messages to explicitly specified recipient(s).
+        Dispatch one-off messages to explicitly specified recipient.
         """
         if messages['subject'] and (messages['body'] or messages['html']):
-            self.send_email_messages(recipient, messages)
+            return self.send_email_messages(recipient, messages)
 
-    def dispatch_order_messages(self, order, messages, event_type=None,
-                                **kwargs):
+    def dispatch_order_messages(self, order, messages, event_type=None, **kwargs):
         """
-        Dispatch order-related messages to the customer
+        Dispatch order-related messages to the customer.
         """
         if order.is_anonymous:
-            if 'email_address' in kwargs:
-                self.send_email_messages(kwargs['email_address'], messages)
-            elif order.guest_email:
-                self.send_email_messages(order.guest_email, messages)
-            else:
-                return
+            email = kwargs.get('email_address', order.guest_email)
+            dispatched_messages = self.dispatch_anonymous_messages(email, messages)
         else:
-            self.dispatch_user_messages(order.user, messages)
+            dispatched_messages = self.dispatch_user_messages(order.user, messages)
 
-        # Create order communications event for audit
-        if event_type is not None:
-            CommunicationEvent._default_manager.create(
-                order=order, event_type=event_type)
+        self.create_communication_event(order, event_type, dispatched_messages)
+
+    def dispatch_anonymous_messages(self, email, messages):
+        dispatched_messages = {}
+        if email:
```

Comment:
I don't think `guest_email` can be `None` - we check set it on the checkout gateway page https://github.com/django-oscar/django-oscar/blob/31ea8dcbcce4ccb6bcbdbb08c3fe9b01103fae2c/src/oscar/apps/checkout/views.py#L80-L82.